speech/grpc/transcribe_streaming_minute.py

```diff
@@ -0,0 +1,294 @@
+# Copyright (C) 2016 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Sample that streams audio to the Google Cloud Speech API via GRPC.
+
+This sample expands on transcribe_streaming.py to work around the 1-minute
+limit on streaming requests. It does this by transcribing normally until
+WRAP_IT_UP_SECS seconds before the 1-minute limit. At that point, it waits for
+the end of an utterance and once it hears it, it closes the current stream and
+opens a new one. It also keeps a buffer of audio around while this is
+happening, that it sends to the new stream in its initial request, to minimize
+losing any speech that occurs while this happens.
+
+Note that you could do this a little more simply by simply re-starting the
+stream after every utterance, though this increases the possibility of audio
+being missed between streams. For learning purposes (and robustness), the more
+complex implementation is shown here.
+"""
+
+from __future__ import division
+
+import argparse
+import collections
+import contextlib
+import functools
+import logging
+import re
+import signal
+import sys
+import time
+
+import google.auth
+import google.auth.transport.grpc
+import google.auth.transport.requests
+import grpc
+from google.cloud.grpc.speech.v1beta1 import cloud_speech_pb2
+from google.rpc import code_pb2
+from grpc.framework.interfaces.face import face
+import pyaudio
+from six.moves import queue
+
+# Seconds you have to wrap up your utterance
+WRAP_IT_UP_SECS = 15
+SECS_OVERLAP = 1
+
+# Audio recording parameters
+RATE = 16000
+CHUNK = int(RATE / 10)  # 100ms
+
+# The Speech API has a streaming limit of 60 seconds of audio*, so keep the
+# connection alive for that long, plus some more to give the API time to figure
+# out the transcription.
+# * https://g.co/cloud/speech/limits#content
+DEADLINE_SECS = 60 * 3 + 5
+SPEECH_SCOPE = 'https://www.googleapis.com/auth/cloud-platform'
+
+
+def make_channel(host, port):
+    """Creates a secure channel with auth credentials from the environment."""
+    # Grab application default credentials from the environment
+    credentials, _ = google.auth.default(scopes=[SPEECH_SCOPE])
+
+    # Create a secure channel using the credentials.
+    http_request = google.auth.transport.requests.Request()
+    target = '{}:{}'.format(host, port)
+
+    return google.auth.transport.grpc.secure_authorized_channel(
+        credentials, http_request, target)
+
+
+def _audio_data_generator(buff, overlap_buffer):
```

Comment:
I feel like we've had this conversation before :) `buffer` turns out to be a reserved word.

https://docs.python.org/2/library/functions.html#buffer